sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -0,0 +1,264 @@
+    else:
+      state.add_state(window, self._element_tag, element)
+      state.add_state(window, self._restriction_tag,
+                      sdf_result.residual_restriction)
+      state.add_state(window, self.watermark_hold_tag,
+                      WatermarkManager.WATERMARK_NEG_INF)
+      # Holding output watermark by setting it to negative infinity.
+      state.set_timer(
+          window, '', TimeDomain.WATERMARK, WatermarkManager.WATERMARK_NEG_INF)
+
+
+class SDFProcessElementInvoker(object):
+  """A utility that requsts checkpoints.
+
+  Requests a checkpoint after processing a predefined number of elements or
+  after a predefined time is elapsed from the process() invocation.
+  """
+
+  class Result(object):
+    def __init__(
+        self, residual_restriction=None, continuation=None,
+        future_output_watermark=None):
+      self.residual_restriction = residual_restriction
+      self.continuation = continuation
+      self.future_output_watermark = future_output_watermark
+
+  def __init__(
+      self, max_num_outputs, max_duration):
+    self._max_num_outputs = max_num_outputs
+    self._max_duration = max_duration
+    self._checkpoint_lock = Lock()
+
+  def test_method(self):
+    raise ValueError
+
+  def invoke_process_element(self, sdf_invoker, element, tracker):
+    assert isinstance(sdf_invoker, DoFnInvoker)
+    assert isinstance(tracker, RestrictionTracker)
+
+    # Using lists for following two variables so that they are accessible from
+    # the closure of the function initiate_checkpoint().
+    checkpointed = []
+    residual_restriction = []
+
+    def initiate_checkpoint():
+      with self._checkpoint_lock:
+        if checkpointed:
+          return
+      residual_restriction.append(tracker.checkpoint())
+      checkpointed.append(object())
+
+    output_processor = _OutputProcessor()
+    Timer(self._max_duration, initiate_checkpoint).start()
+    sdf_invoker.invoke_process(
+        element, restriction_tracker=tracker, output_processor=output_processor)
+
+    assert output_processor.output_iter is not None
+    output_count = 0
+
+    # We have to expand and re-yield here to support ending execution for a
+    # given number of output elements as well as to capture the
+    # ProcessContinuation of one was returned.
+    for output in output_processor.output_iter:
+      if isinstance(output, ProcessContinuation):
+        # Taking a checkpoint so that we can determine primary and residual
+        # restrictions.
+        initiate_checkpoint()
+
+        # A ProcessContinuation should always be the last element produced by
+        # the output iterator.
+        # TODO: support continuing after the specified amount of delay.
+        break
+
+      yield output
+      output_count += 1
+      if self._max_num_outputs and output_count >= self._max_num_outputs:
+        initiate_checkpoint()
+
+    tracker.check_done()
+    result = (
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:0-->
Can you do something like:

```python
residual_restriction = residual_restrictions[0] if residual_restrictions else None
yield SDFProcessElementInvoker.Result(residual_restriction=residual_restriction)
```

amd rename `residual_restriction` to `residual_restrictions` above?